proposals/custom-states-and-state-pseudo-class.md

```diff
@@ -0,0 +1,105 @@
+# `ElementInternals.states` and the `:state()` pseudo class
+
+Author: @rakina, @domenic
+
+## Introduction
+
+Built-in elements have certain “states” that can change over time depending on user interaction and other factors, and are exposed to web authors through pseudo classes. For example, some form controls have the “invalid” state, which is exposed through the `:invalid` pseudo class.
+
+Like built-in elements, custom elements can have various states to be in too, and web authors might want to expose these states in a similar fashion as the built-in elements. With the proposed `ElementInternals.state`  property, custom element authors can add and modify custom states for the custom elements, and allow them to be selected with the `:state()` selector.
+
+### Goals
+
+-   Give a way for custom element authors to expose varying states of the custom elements to custom element users
+-   Allow styling of custom elements that differ according state
+-   Allow styling of custom elements that differ according to state even when they are only accessible via shadow parts. That is, `element::part(x):state(y)` should work if `::part(x)` selects a custom element with state y..
+    
+
+### Non-goals
+
+-   Allowing adding custom states to built-in elements. States reflect outward what the element implementation author wishes to show the world; for application-level state, use the `class=""` attribute or similar.
+-   Allowing custom element authors to modify already-existing/built-in states/pseudo classes, e.g. `:checked`. (This might be added in the future, though.)
+
+## Example
+
+```html	
+<!DOCTYPE html>
+<body>
+<script>
+class LabeledCheckbox extends  HTMLElement {
+  constructor() {
+    super();
+    this._internals = this.attachInternals();
+    this._checked = false;
+    this.addEventListener('click', this._onClick.bind(this));
+
+    this._shadowRoot = this.attachShadow({mode: "open"});
+    this._shadowRoot.innerHTML =
+      "<style>\
+        :host::before { content: '[]'; }\
+        :host(:state(checked))::before { content: '[x]' }\
+      </style>\
+      <span><slot>Label</slot></span>";
+  }
+  
+  get checked() { return this._checked; }
+  
+  set checked(flag) {
+    this._checked = !!flag;
+    if (this._checked) {
+      this._internals.states.add("checked");
+    } else {
+      this._internals.states.remove("checked");
+    }
+  }
+
+  _onClick(event) {
+    this.checked = !this._checked;
+  }
+  
+}
+  
+class QuestionBox extends HTMLElement {
+  constructor() {
+    super();
+    this._shadowRoot = this.attachShadow({mode: "open"});
+    this._shadowRoot.innerHTML =
+      "<div><slot>Do you agree?</slot></div>\
+      <labeled-checkbox part='checkbox'>Yes</labeled-checkbox>";
+  }
+}
+customElements.define('labeled-checkbox', LabeledCheckbox);
+customElements.define('question-box', QuestionBox);
+</script>
+
+<style>
```

Comment:
It's probably worth adding an example of `:state()` without `::part()`, e.g. by using `labeled-checkbox` by itself.